src/Nethermind/Nethermind.Core/Threading/DStack.cs

```diff
@@ -0,0 +1,146 @@
+// SPDX-FileCopyrightText: 2025 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace Nethermind.Core.Threading;
+
+/// <summary>
+/// Like a concurrent stack, but also support dequeue.
+/// It will not reduce buffer size on dequeue/pop.
+/// The dequeue may return false and not return item if blocked.
+/// </summary>
+/// <param name="initialCapacity"></param>
+/// <typeparam name="T"></typeparam>
+public class DStack<T>(int initialCapacity)
+{
+    private SpinLock _locker = new SpinLock();
```

Comment:
@benaadams can this be made lockless?

Reply:
Which one is this implementation? https://github.com/crossbeam-rs/crossbeam/blob/master/crossbeam-deque/src/deque.rs